### PR TITLE
Move the CI to github-actions

### PR DESCRIPTION
**Pull request by pierreglaser**
This PR changes the continuous integration provider of cloudpickle from `travis` to the in-house `github` provider available via `github-actions`

### CODE REVIEW

**Comment by pierreglaser** [comment]
just rebasing with some branches that fix various issues with the current test suite.

**Comment by pierreglaser** [comment]
I can't check b/c I don't have enough rights: can you tell me if we already have a `CODECOV_TOKEN` under the `Secrets` sections of the repository settings?

**Comment by ogrisel** [comment]
Let's convert the coverage report to the xml format as we do on the loky CI, this way we can use the language agnostic codecov bash upload script that does not need `CODECOV_TOKEN` if I recall correctly.

https://github.com/joblib/loky/blob/master/tox.ini#L31

https://github.com/joblib/loky/blob/master/.azure_pipeline.yml#L97-L100

**Comment by pierreglaser** [comment]
The good news is that `test_local_tls` is the only failing `distributed` test, the rest of the test suite runs fine.

File: .github/workflows/pythonapp.yml [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        $PYTHON_EXE -m pip install -e .
+        $PYTHON_EXE -m pip install -r dev-requirements.txt
+        $PYTHON_EXE ci/install_coverage_subprocess_pth.py
+    - name: Display Python version
+      run: $PYTHON_EXE -c "import sys; print(sys.version)"
+    - name: Run linting
+      run: |
+        # stop the build if there are Python syntax errors or undefined names
+        $PYTHON_EXE -m flake8 . --count --verbose --select=E901,E999,F821,F822,F823 \
+          --show-source --statistics
```
[the remark this answers is not in the document]
Comment (pierreglaser) [reply .github/workflows/pythonapp.yml]: I don't know if this remark was linked to https://github.com/cloudpipe/cloudpickle/pull/324#discussion_r371264933, but if any `github-action` job fails, then all other running jobs gets cancelled. So if linting fails ona job of its own (and not all entries of the matrix), it is enough for the CI to stop.

File: .github/workflows/testing.yml [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+on: [push]
+
+jobs:
+  build:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest]
+        python_version: [2.7, 3.5, 3.6, 3.7, 3.8]
+        exclude:
+          # do not test each porsisbl (os, python version) configuration
```
Comment (ogrisel) [inline .github/workflows/testing.yml]:
```suggestion
          # do not test each possible (os, python version) configuration
```

File: .github/workflows/testing.yml [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+          - os: windows-latest
+            python_version: 3.6
+          - os: windows-latest
+            python_version: 3.8
+          - os: macos-latest
+            python_version: 3.5
+          - os: macos-latest
+            python_version: 3.6
+          - os: macos-latest
+            python_version: 3.8
```
Comment (ogrisel) [inline .github/workflows/testing.yml]: Why are those excluded? In particular I think we should test the latest python version (3.8) for all the supported platforms.

Skipping 3.6 or 3.7 on some platforms to speed up CI is fine but I would rather always test the boundaries (2.7, 3.5 and 3.8 at the moment) on all platforms.

File: .github/workflows/testing.yml [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    - name: Set up Python 3.7
+      uses: actions/setup-python@v1
+      with:
+        python-version: ${{ matrix.python_version }}
+    - name: Install project and dependencies
+      run: |
+        bash ./.github/scripts/install_downstream_project.sh
+    - name: Test the downstream project
+      run: |
+        bash ./.github/scripts/test_downstream_project.sh
```
Comment (ogrisel) [inline .github/workflows/testing.yml]: nitpick: please insert a blank line after this job block.

File: .github/workflows/testing.yml (hunk 3 above)
Comment (pierreglaser) [reply .github/workflows/testing.yml]: Yes I just mimicated the entries we had for `windows`. You are right, we might as well test the boundaries.

**Comment by pierreglaser** [comment]
Regarding the `test_local_tls`, I proposed an approved fix in dask/distributed#3418, that has yet to be merged.

**Comment by pierreglaser** [comment]
@ogrisel do you want me to run work on a `flake8_diff` workflow here before merging? I started extracting the gist of https://github.com/scikit-learn/scikit-learn/blob/837d2b55522783174291ea5cd6b771e11583c270/build_tools/circle/linting.sh, but I am not done yet.

**Comment by pierreglaser** [comment]
`joblib` errors out with a core dump...

**Comment by pierreglaser** [comment]
Apparently we have another distributed failure. And it seems to be related to dask/distributed#3418.

**Comment by pierreglaser** [comment]
the `joblib` failure is a benign failure due a change in the brand new `2.10` distributed version. I am looking in to the distributed failure now.

**Comment by ogrisel** [comment]
Let's skip the distributed tests that are too hard to debug but pass on distributed's travis (as long as the issue is properly reported upstream).

File: .github/workflows/testing.yml [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        os: [ubuntu-latest, windows-latest, macos-latest]
+        python_version: [2.7, 3.5, 3.6, 3.7, 3.8, "pypy3"]
+        exclude:
+          # Do not test all minor version on all platforms, especially if they
+          # are not the oldest/newest supported versions
+          - os: windows-latest
+            python_version: 3.6
+          - os: windows-latest
+            python_version: 3.7
+          - os: windows-latest
```
Comment (pierreglaser) [inline .github/workflows/testing.yml]: I am failing to build some dependencies on`pypy` and `Windows` - this is why I am skipping this config. I should add a comment.

**Comment by pierreglaser** [comment]
OK -- the PR is good for merging. To make it green we need to merge joblib/joblib#1003

**Comment by pierreglaser** [comment]
CI is all green. Merging!

**Pull request merged**